File: src/PgmStudio.Client/wwwroot/js/studio/shared/game-colors.js
/** PGM / Minecraft color palettes. */

export const PGM_CHAT_COLORS = [
  { value: "black",        label: "Black",        hex: "#000000" },
  { value: "dark blue",    label: "Dark Blue",    hex: "#0000AA" },
  { value: "dark green",   label: "Dark Green",   hex: "#00AA00" },
  { value: "dark aqua",    label: "Dark Aqua",    hex: "#00AAAA" },
  { value: "dark red",     label: "Dark Red",     hex: "#AA0000" },
  { value: "dark purple",  label: "Dark Purple",  hex: "#AA00AA" },
  { value: "gold",         label: "Gold",         hex: "#FFAA00" },
  { value: "gray",         label: "Gray",         hex: "#AAAAAA" },
  { value: "dark gray",    label: "Dark Gray",    hex: "#555555" },
  { value: "blue",         label: "Blue",         hex: "#5555FF" },
  { value: "green",        label: "Green",        hex: "#55FF55" },
  { value: "aqua",         label: "Aqua",         hex: "#55FFFF" },
  { value: "red",          label: "Red",          hex: "#FF5555" },
  { value: "light purple", label: "Light Purple", hex: "#FF55FF" },
  { value: "yellow",       label: "Yellow",       hex: "#FFFF55" },
  { value: "white",        label: "White",        hex: "#FFFFFF" },
];

export const MINECRAFT_DYE_COLORS = [
  { value: "white",      label: "White",      hex: "#FFFFFF" },
  { value: "orange",     label: "Orange",     hex: "#D87F33" },
  { value: "magenta",    label: "Magenta",    hex: "#B24CD8" },
  { value: "light blue", label: "Light Blue", hex: "#6699D8" },
  { value: "yellow",     label: "Yellow",     hex: "#E5E533" },
  { value: "lime",       label: "Lime",       hex: "#7FCC19" },
  { value: "pink",       label: "Pink",       hex: "#F27FA5" },
  { value: "gray",       label: "Gray",       hex: "#4C4C4C" },
  { value: "silver",     label: "Silver",     hex: "#999999" },
  { value: "cyan",       label: "Cyan",       hex: "#4C7F99" },
  { value: "purple",     label: "Purple",     hex: "#7F3FB2" },
  { value: "blue",       label: "Blue",       hex: "#334CB2" },
  { value: "brown",      label: "Brown",      hex: "#664C33" },
  { value: "green",      label: "Green",      hex: "#667F33" },
  { value: "red",        label: "Red",        hex: "#993333" },
  { value: "black",      label: "Black",      hex: "#191919" },
];

export function chatColorHex(name) {
  const n = (name ?? "").replace(/_/g, " ").toLowerCase();
  return PGM_CHAT_COLORS.find(c => c.value === n)?.hex ?? "#475569";
}

const _normDye = name => (name ?? "").replace(/_/g, " ").toLowerCase();

export function dyeColorHex(name) {
  return MINECRAFT_DYE_COLORS.find(c => c.value === _normDye(name))?.hex ?? "#475569";
}

export function dyeColorLabel(name) {
  return MINECRAFT_DYE_COLORS.find(c => c.value === _normDye(name))?.label
    ?? name.replace(/_/g, " ");
}

// ── Team colour assignment ──────────────────────────────────────────────────

// Preferred order for auto-assigning a new team's colour: the bright, readily
// distinguishable colours first (typical CTW uses red/blue, then green/yellow…).
export const TEAM_COLOR_PRIORITY = [
  "red", "blue", "green", "yellow", "aqua", "gold", "light purple", "dark purple",
  "dark aqua", "dark green", "dark red", "dark blue", "gray", "dark gray", "white", "black",
];

const _normChat = name => (name ?? "").replace(/_/g, " ").toLowerCase();

/** Pick the next unused team colour in priority order, or null when all 16 are
 *  taken. `usedColors` is the existing teams' colour values (any case / `_`-form). */
export function nextTeamColor(usedColors = []) {
  const used = new Set([...usedColors].map(_normChat));
  for (const value of TEAM_COLOR_PRIORITY) {
    if (!used.has(value)) return PGM_CHAT_COLORS.find(c => c.value === value);
  }
  return null;
}

File: src/PgmStudio.Client/Models/GameColors.cs
using System.Text.Json;

namespace PgmStudio.Client.Models;

/// <summary>
/// Minecraft 1.8 colour tables from Bukkit, loaded from the shared <c>game-colors.json</c> — the same file the
/// JS side reads (<c>render/palette.js</c>), so the two copies cannot drift. <see cref="ChatColors"/> is
/// <c>org.bukkit.ChatColor</c> (chat / formatting colours), <see cref="DyeColors"/> is <c>org.bukkit.DyeColor</c>
/// (dye / wool colours, indexed by data value); the hex is display-only, the value is what the wire carries.
/// Team-colour assignment is a studio convenience, not a Bukkit thing, so its order lives here in code.
/// </summary>
public static class GameColors
{
    public readonly record struct Color(string Value, string Label, string Hex);

    /// <summary><c>org.bukkit.ChatColor</c>, in enum-ordinal order (black 0 … white 15).</summary>
    public static readonly Color[] ChatColors;

    /// <summary><c>org.bukkit.DyeColor</c>, indexed by data value (white 0 … black 15).</summary>
    public static readonly Color[] DyeColors;

    static GameColors()
    {
        var assembly = typeof(GameColors).Assembly;
        var resource = assembly.GetManifestResourceNames()
            .First(n => n.EndsWith("game-colors.json", StringComparison.Ordinal));
        using var stream = assembly.GetManifestResourceStream(resource)!;
        using var doc = JsonDocument.Parse(stream);
        ChatColors = ReadTable(doc.RootElement.GetProperty("chat"));
        DyeColors = ReadTable(doc.RootElement.GetProperty("dye"));
    }

    // Parse via JsonDocument rather than a typed deserialize: the WASM app disables reflection-based
    // serialization, and this keeps the load working without a source-generated context for one small file.
    private static Color[] ReadTable(JsonElement array)
    {
        var colors = new List<Color>(array.GetArrayLength());
        foreach (var entry in array.EnumerateArray())
            colors.Add(new Color(
                entry.GetProperty("value").GetString()!,
                entry.GetProperty("label").GetString()!,
                entry.GetProperty("hex").GetString()!));
        return [.. colors];
    }

    // Auto-assign order for a new team's colour (bright, distinguishable first).
    private static readonly string[] TeamColorPriority =
    [
        "red", "blue", "green", "yellow", "aqua", "gold", "light purple", "dark purple",
        "dark aqua", "dark green", "dark red", "dark blue", "gray", "dark gray", "white", "black",
    ];

    private static string Norm(string? name) => (name ?? "").Replace('_', ' ').ToLowerInvariant();

    public static string ChatHex(string? name)
        => ChatColors.FirstOrDefault(c => c.Value == Norm(name)).Hex is { Length: > 0 } h ? h : "#475569";

    public static string DyeHex(string? name)
        => DyeColors.FirstOrDefault(c => c.Value == Norm(name)).Hex is { Length: > 0 } h ? h : "#475569";

    /// <summary>Next unused team colour in priority order, or null when all 16 are taken.</summary>
    public static Color? NextTeamColor(IEnumerable<string?> usedColors)
    {
        var used = usedColors.Select(Norm).ToHashSet();
        foreach (var value in TeamColorPriority)
            if (!used.Contains(value)) return ChatColors.First(c => c.Value == value);
        return null;
    }

    /// <summary>The first <paramref name="n"/> team colours in priority order (red, blue, green, …).</summary>
    public static IReadOnlyList<Color> FirstTeamColors(int n)
    {
        var result = new List<Color>();
        var used = new List<string?>();
        for (var i = 0; i < n && NextTeamColor(used) is { } c; i++) { result.Add(c); used.Add(c.Value); }
        return result;
    }
}
